### PR TITLE
Add and fix

### PR DESCRIPTION
**Pull request by Frouk3**
Fix logo sequence and add Infinite Wig B fix

### CODE REVIEW

**Comment by Frouk3** [comment]
hey @CookiePLMonster

**Comment by CookiePLMonster** [comment]
Hey,

I barely know the game as-is so I can't comment on the fix itself and I can only assume it's OK. However, in the current state I cannot accept it for 2 reasons:
1. The entire file got reformatted.
2. I'm not sure why did you change patterns to hardcoded addresses. I don't know how many different game executables are in the wild and so as a rule I always stick to patterns, even if there is just one commonly used executable.

**Comment by Frouk3** [comment]
> Hey,
> 
> I barely know the game as-is so I can't comment on the fix itself and I can only assume it's OK. However, in the current state I cannot accept it for 2 reasons:
> 
> 1. The entire file got reformatted.
> 2. I'm not sure why did you change patterns to hardcoded addresses. I don't know how many different game executables are in the wild and so as a rule I always stick to patterns, even if there is just one commonly used executable.

1. File got reformatted because of Visual Studio didn't update the main file, and i used Ctrl+C and Ctrl+V in github
2. I changed some because of patterns weren't finding a correct address with array of bytes, anything but skip logo sequence and fixed wig b, weren't changed

**Comment by Frouk3** [comment]
maybe non-related:
Game has dynamic base address(it may change on game restart)
I'm making ida database, and probably it'll be long to fill it up

**Comment by CookiePLMonster** [comment]
> Game has dynamic base address(it may change on game restart)

With patterns this doesn't matter either.

**Comment by Frouk3** [comment]
but they may not work always

**Comment by CookiePLMonster** [comment]
They do, that's why I always use them - long as you don't use data addresses in patterns. The ones produced by IDA's SigMaker are always reliable.

**Review by CookiePLMonster — CHANGES_REQUESTED** [review]
Any update on this? Like I said, I am willing to accept those changes but not if they don't use patterns and reformat the code.

**Comment by Frouk3** [comment]
> Any update on this? Like I said, I am willing to accept those changes but not if they don't use patterns and reformat the code.

can you rephrase your sentence? I cannot understand it? Am i allowed to add without patterns?

**Comment by CookiePLMonster** [comment]
> Am i allowed to add without patterns?

No - please make this change use patterns and remove code formatting changes.

**Comment by Frouk3** [comment]
> > Am i allowed to add without patterns?
> 
> No - please make this change use patterns and remove code formatting changes.

I'll try

**Comment by Frouk3** [comment]
how to use patterns? I don't understand offset from pattern

**Comment by Frouk3** [comment]
i'll leave patterns, so you will check the offset from them:
Skip logo sequence: 85 C9 74 13 E8 BF 6E FF FF
WIG B Disable on blade mode: 75 14 F7 05 94 A0 ?? 01
WIG B Enable in blade mode: 75 0E 8B 16 8B 82 2C 03 00 00 FF D0

**Comment by CookiePLMonster** [comment]
> how to use patterns? I don't understand offset from pattern

`.get()` takes an offset from the first byte found in the pattern, see the intro skip pattern you removed for a clear example on how it works.

**Comment by Frouk3** [comment]
> > how to use patterns? I don't understand offset from pattern
> 
> `.get()` takes an offset from the first byte found in the pattern, see the intro skip pattern you removed for a clear example on how it works.

so it would look like this?
02489C: 85 C9
Offset to jmp is 2
02489E: 74 13

**Comment by Frouk3** [comment]
@CookiePLMonster

**Comment by CookiePLMonster** [comment]
Thanks, I'll review it shortly. Regarding this fix, what does it do exactly, so I know what to put in a changelog?

**Comment by Frouk3** [comment]
> Thanks, I'll review it shortly. Regarding this fix, what does it do exactly, so I know what to put in a changelog?

OK.
A fix of skip logo sequence is just changed pattern of bytes, where original pattern wasn't founding anything. When launched with SkipLogoSequences setted to 1. It wasn't doing anything. Nothing was occuring but crash. Fixed infinite wig b is allowing player to stay in ripper mode when blade mode is occured. First check i've done is fixed ripper mode when enabling blade mode, second is allowing to enable ripper mode in blade mode(Game does allow this without a fix, but without infinite wig b, with fix, it does the same, but allows ripper mode with wig b)

**Comment by Frouk3** [comment]
@CookiePLMonster ?

**Comment by CookiePLMonster** [comment]
Sorry, I've not forgotten about that - I am busy with a contract right now.

**Review by CookiePLMonster — CHANGES_REQUESTED** [review]
Just one more note:

I changed the Wig B patterns to have no data dependencies, they are functionally equivalent otherwise. However, I don't understand the need to change the intro skip pattern - what I had in the code already is unambiguous already:

image

I suspect you may have a modified EXE and therefore the pattern fails. If that is the case, I wouldn't worry about the crash - since the time I prototyped this SilentPatch, ModUtils and patterns evolved considerably so before releasing I will migrate this codebase to use transactional patterns - which means that in the case of a match failure, only this one fix is going to be skipped, and everything before/after it is going to work fine.

I also took the liberty to undo your changes partially (already pushed to your branch) to keep the diff minimal. Can you check the intro pattern again? Once that is cleared, I'm going to squash all commits and merge.

EDIT:
For posterity - my game executable is 26 347 008 bytes, SHA-1 hash `ad6750825e2158a9e5cbaf31f90d23c9e4948746`.

**Comment by Frouk3** [comment]
i changed pattern, and didn't know that there's mod installed that allows to skip the intro

i have the same SHA-1 hash as yours

**Comment by Frouk3** [comment]
image
i guess that mod that i have installed is replaces `v5 = this[35];`
and that's were the confusion

**Comment by Frouk3** [comment]
in case, i should make `if (!showLogoSequence) showLogoSequence = pattern( skipIntroPatternOnFail ).get_one();`?

**Comment by Frouk3** [comment]
and i have two questions, on how many percent you have database filled up, and the second one, how do i install library of Visual Studion 2017?

**Comment by CookiePLMonster** [comment]
> in case, i should make `if (!showLogoSequence) showLogoSequence = pattern( skipIntroPatternOnFail ).get_one();`?

Just leave it as-is and I will revert the code to how it was before your changes.



> and i have two questions, on how many percent you have database filled up, and the second one, how do i install library of Visual Studion 2017?

I don't have a DB for MGR at all, and you **can** upgrade your project to VS2022 locally like you did - just don't commit your `.vcxproj` changes and have them local-only.

**Review by CookiePLMonster — APPROVED** [review]

**Comment by CookiePLMonster** [comment]
Should be good to go now.

**Comment by Frouk3** [comment]
> > in case, i should make `if (!showLogoSequence) showLogoSequence = pattern( skipIntroPatternOnFail ).get_one();`?
> 
> Just leave it as-is and I will revert the code to how it was before your changes.
> 
> > and i have two questions, on how many percent you have database filled up, and the second one, how do i install library of Visual Studion 2017?
> 
> I don't have a DB for MGR at all, and you **can** upgrade your project to VS2022 locally like you did - just don't commit your `.vcxproj` changes and have them local-only.

in case, i'm filling up database

**Comment by CookiePLMonster** [comment]
Thanks, merged now. I'll have to get this code up to 2023 standards a bit, then port the frame pacing fixes from Wonderful 101 and it should be good to go.
